### PR TITLE
Make (user) instance aware of delegated cgroup controllers

### PR DESCRIPTION
**Pull request by Werkov**
This series resolves the warnings that are reported in #17862 and #18047. Normally, everything works fine in the host instance (:= PID 1 in init namespace), however, in "nested" cases not all controllers may be available:

  - v1 non-privileged user instance: we don't allow delegation,
  - v1 container: the cgroup hierarch(y|ies) may be mounted read-only,
  - v2: not all controllers are necessarily delegated.

The current approach was to consider all controllers "reported by the kernel" as available and if any of the cases above were hit during cgroup realization, silently accept the error. (Well not so silently in v247.)

This series makes the systemd instance initially probe what controllers are available to it and then all operations are restricted just to this set preventing vain attempts at non-delegated controllers.

Why is this posted as RFC? This change adds one more step when new controllers are delegated to the user instance (this is relevant to v2). It wouldn't _just work_ but the user instance has to update its `manager->cgroup_supported`, currently via `systemctl --user daemon-reexec`. Neither behavior is documented so there should be relative freedom to change it. I'd argue that the latter is more deterministic, however, I'd like to hear if there are any reasons for the former. Also if there's inclination towards the two-step delegation, I'd modify this to apply also with `systemctl --user  daemon-reload`.

An alternative to all of this would be to (again) silence the expected errors from v1 hierarchies operations.

### CODE REVIEW

**Comment by Werkov** [comment]
The tests finished, let me cc: @benzea @cdown @DaanDeMeyer Thanks.

**Comment by benzea** [comment]
Personally, I don't think it is a big issue that one needs to reexec the user instance to update the controllers.

I do wonder though if the user instance should print a warning (once) if you try to use a controller that is supported by the host but unusable because it has not been delegated.

File: src/basic/cgroup-util.c [hunk 1]
```diff
@@ -527,15 +527,21 @@ int cg_get_path(const char *controller, const char *path, const char *suffix, ch
...
-        cc = strjoina("/sys/fs/cgroup/", dn);
-
-        if (laccess(cc, F_OK) < 0)
+        cpath = strjoina("/sys/fs/cgroup/", dn);
+        if (root)
+                /* Also check that:
+                 * - possible subcgroup is created at root,
+                 * - we can modify the hierarchy.
+                 * "Leak" cpath on stack */
+                cpath = strjoina(cpath, root, "/cgroup.procs");
```
Comment (keszybz) [inline src/basic/cgroup-util.c:542]: Maybe `cpath = strjoina("/sys/fs/cgroup/", dn, root, root ? "/cgroup.procs" : NULL)` ?

**Review by keszybz — COMMENTED** [review]
Yeah, I think this is a good approach. The daemon-reexec step isn't a big issue: we don't expect or need the user manager to react dynamically to controllers being delegated or undelegated. Changing this dynamically isn't really something we can support. But it matters that we don't emit warnigs when not appropriate.

File: src/basic/cgroup-util.c [hunk 2]
```diff
@@ -572,10 +547,23 @@ int cg_get_path_and_check(const char *controller, const char *path, const char *
...
+                return -EINVAL;
+
+        r = cg_all_unified();
         if (r < 0)
                 return r;
+        if (r > 0) {
+                /* In the unified hierarchy all controllers are considered accessible,
+                 * except for the named hierarchies */
+                if (startswith(controller, "name="))
+                        return -EOPNOTSUPP;
```
Comment (poettering) [inline src/basic/cgroup-util.c:566]: this dropped the logic of allowing `name=systemd` in the v2 case. On purpose?

**Review by poettering — COMMENTED** [review]
patch looks good to me, but see one comment

File: src/basic/cgroup-util.c (hunk 1 above)
Comment (Werkov) [reply src/basic/cgroup-util.c:542]: It's a bit tricky (the middle NULL which is also the last) but it does the job. Thanks for the fixup (e4645ca5995551772f240407c4eca970f1df65c7).

File: src/basic/cgroup-util.c (hunk 2 above)
Comment (Werkov) [reply src/basic/cgroup-util.c:566]: The reasoning is that in the unified mode, we work only with the v2 hierarchy and that doesn't allow/need any `name=`. So `name=systemd` should be bathwater too, not a baby.
The logic should be same as before the change, the relevant check was against 
> #define SYSTEMD_CGROUP_CONTROLLER "_systemd"

that itself doesn't start with `name=`.

@poettering Does that clarify?

**Pull request merged**